challenges/cmojidra_Headstorm/cmojidraDatabase/Headstorm-DB/src/main/java/MigrateData.java

```diff
@@ -0,0 +1,145 @@
+        System.out.println("CREATE TABLE QUERIES ARE: ");
+        System.out.println();
+        /**
+         * Creating SQL tables
+         */
+        createTables();
+
+        /**
+         * Reading input json file and parsing the json object
+         */
+        JSONParser jsonParser = new JSONParser();
+
+        try {
+            FileReader reader = new FileReader("inputfile.txt");
+            Object obj = jsonParser.parse(reader);
+
+            JSONArray records = (JSONArray) obj;
+            System.out.println();
+            System.out.println("INSERT QUERIES ARE: ");
+            System.out.println();
+            for(int i = 0; i < records.size(); i++){
+                JSONObject jsonObject = (JSONObject)records.get(i);
+               migrateData.parseRecords(jsonObject);
+            }
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (ParseException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+
+    }
+
+    /**
+     * Method to parse each record and generate insert queries
+     */
+    private void parseRecords(JSONObject record)
+    {
+
+        long record_id = (long) record.get("record_id");
+        String name = (String) record.get("name");
+        String cell_phone = (String) record.get("cell_phone");
+        String work_phone = (String) record.get("work_phone");
+        String email = (String) record.get("email");
+        String address = (String) record.get("address");
+        long basic_widget_order = (long) record.get("basic_widget_order");
+        long advanced_widget_order = (long) record.get("advanced_widget_order");
+        String protection_plan = (String) record.get("protection_plan");
+        long repeated_user;
+        String insertBasic="";
+        String insertAdvanced="";
+        /**
+         * Checking if the user exists or not
+         */
+        if (!map.containsKey(email)){
+            String insertUsers = "INSERT into users values(" + user_id + ", \"" + name + "\", \"" + work_phone+"\", \"" +
+                    cell_phone+ "\", \"" + email + "\", \"" + address + "\", " + protection_plan +" );";
+            System.out.println(insertUsers);
+            map.put(email,user_id);
+            insertBasic = "INSERT into basic_orders values(" + basic_order_id + ", " + user_id + ", " + basic_widget_order + " );";
+            insertAdvanced = "INSERT into advanced_orders values(" + advanced_order_id + ", " + user_id + ", " + advanced_widget_order +" );";
+            user_id++;
+        }
+        else{
+            repeated_user = map.get(email);
+            insertBasic = "INSERT into basic_orders values(" + basic_order_id + ", " + repeated_user + ", " + basic_widget_order + " );";
+            insertAdvanced = "INSERT into advanced_orders values(" + advanced_order_id + ", " + repeated_user+", " + advanced_widget_order + " );";
+        }
```

Comment:
What would happen if we had two separate users with the same email in the NOSQL data store?

Reply:
Based on my assumption of the real-world scenarios, an email id can be associated with only 1 account for 1 user. So if a user tries to create an account with an already registered email ID, he'll be prompted to login since the email id is already registered. Nevertheless, I totally get your point of using something else as a primary key. This makes sense since it gives us complete control of handling the primary keys and not depend on a value that is being generated by an external entity (Gmail, Microsoft, Yahoo).

Reply:
I definitely agree that we should limit our users to having unique email addresses. Most likely this is the way that the legacy NOSQL store worked. That said, if we are making that assumption we would definitely want to confirm it is valid before we ran this code. If not we might end up attaching orders to incorrect users and dropping duplicate users.

If we still wanted to have an auto-incrementing primary key for our users, but also make sure our data integrity is valid (i.e. no duplicate emails) how would we do that?

Reply:
Before migrating we can check, using aggregate functions on the NoSQL database, whether multiple documents have the same email ids. If there are no documents with duplicate email ids then in that case we are free to continue with the migration. However, if there are records with the same email ids then we can send a notification to that email id with an option of choosing a primary account. Based on the selection we will retain the corresponding document and merge the orders of the other duplicate accounts to the chosen one. This has been handled using the hashmap in the code. This scenario won't happen in an ideal scenario as it has been the standard norm in the industry for a long time that one email id can belong to only one account.

To handle this scenario, in my code I have created a user_id that is an auto-generated key. This user_id will be assigned if the email id appears for the first time in the document store. If the document with the same email id occurs again then that won't be stored in the user table. For handling, this one hashmap can be created with email id as key and user_id as value. If the same email id comes again then the old user_id will be retrieved instead of assigning a new email_id. 

Hence, this scenario has already been handled in the code submitted by me.

Reply:
Yeah checking the data in the NoSql store using aggregate beforehand would definitely be a good idea - nice!